cmd/exec/exec.go

```diff
@@ -0,0 +1,170 @@
+		IOFS: &afero.IOFS{Fs: fs},
+	}
+
+	tmp.Flags().SortFlags = false
+
+	tmp.Flags().BoolVarP(&listTools, "list", "l", false, "list tools")
+	err := tmp.RegisterFlagCompletionFunc("list", flagCompletion)
+	cobra.CheckErr(err)
+
+	// tmp.Flags().StringVarP(&selectedToolInfo, "info", "i", "", "display the selected template's configuration and default values")
+	// err = tmp.RegisterFlagCompletionFunc("info", flagCompletion)
+	// cobra.CheckErr(err)
+
+	tmp.Flags().StringVar(&format, "format", "table", "display output in table or json format")
+	err = tmp.RegisterFlagCompletionFunc("format", func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+		if len(args) != 0 {
+			return nil, cobra.ShellCompDirectiveNoFileComp
+		}
+		var formats = []string{"table", "json"}
+		return utils.Find(formats, toComplete), cobra.ShellCompDirectiveNoSpace | cobra.ShellCompDirectiveNoFileComp
+	})
+	cobra.CheckErr(err)
+
+	tmp.Flags().StringVar(&localToolPath, "toolpath", "", "location of installed tools")
+	err = viper.BindPFlag("toolpath", tmp.Flags().Lookup("toolpath"))
+	cobra.CheckErr(err)
+
+	return tmp
+}
+
+func preExecute(cmd *cobra.Command, args []string) error {
+	defaultTemplatePath, err := utils.GetDefaultToolPath()
+	if err != nil {
+		return err
+	}
+
+	viper.SetDefault("toolpath", defaultTemplatePath)
+	localToolPath = viper.GetString("toolpath")
+
+	cachedTools = prmApi.List(localToolPath, "")
+
+	return nil
+}
+
+func validateArgCount(cmd *cobra.Command, args []string) error {
+	// show available tools if user runs `prm exec`
+	if len(args) == 0 && !listTools {
+		listTools = true
+	}
+
+	if len(args) >= 1 {
+		if len(strings.Split(args[0], "/")) != 2 {
+			return fmt.Errorf("Selected tool must be in AUTHOR/ID format")
+		}
+		selectedTool = args[0]
+	}
+
+	return nil
+}
+
+func flagCompletion(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	if localToolPath == "" {
+		err := preExecute(cmd, args)
+		if err != nil {
+			log.Error().Msgf("Unable to set tool path: %s", err.Error())
+			return nil, cobra.ShellCompDirectiveError
+		}
+	}
+	if len(args) != 0 {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+	localToolPath = viper.GetString("toolpath")
+
+	return completeName(localToolPath, toComplete), cobra.ShellCompDirectiveNoSpace | cobra.ShellCompDirectiveNoFileComp
+}
+
+func completeName(cache string, match string) []string {
+	var names []string
+	for _, tool := range cachedTools {
+		namespacedTemplate := fmt.Sprintf("%s/%s", tool.Plugin.Author, tool.Plugin.Id)
+		if strings.HasPrefix(namespacedTemplate, match) {
+			m := namespacedTemplate + "\t" + tool.Plugin.Display
+			names = append(names, m)
+		}
+	}
+	return names
+}
+
+func execute(cmd *cobra.Command, args []string) error {
+	span := telemetry.GetSpanFromContext(cmd.Context())
+	// Add tool to span if needed
+	if len(args) == 1 {
+		telemetry.AddStringSpanAttribute(span, "tool", args[0])
+	}
```

Comment:
Asking for clarity more than anything, in case I'm missing something from #15 - from prior discussions, it seemed like this was unnecessary as we're capturing all subcommands and their args from the root command level [here](https://github.com/puppetlabs/prm/blob/main/main.go#L33-L34). 

https://github.com/puppetlabs/prm/blob/387f13d787211fa53945b225d952620ce60fcb00/cmd/root/root.go#L94-L110

Reply:
@michaeltlombardi can confirm on this?

Reply:
We always get back the full arguments, but it is useful to be able to filter on tool as well (instead of having to do an argument matcher)

See: https://ui.honeycomb.io/puppet-modules/datasets/pct_dev/result/JDDnoFdo34o/trace/hrV2s4nn3a8?span=8ec405b833da1b47

Reply:
![image](https://user-images.githubusercontent.com/14190564/144094903-ad093d9f-c510-4514-851a-7edf358c1ff2.png)